clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -338,6 +338,18 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
     if (type_is_instr(memref.instr.type) ||
         memref.instr.type == TRACE_TYPE_PREFETCH_INSTR ||
         memref.instr.type == TRACE_TYPE_INSTR_NO_FETCH) {
+        // We do not bother to support legacy traces without encodings.
+        bool has_encoding = TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, shard->file_type_);
+        addr_t trace_pc = shard->prev_instr_.instr.addr;
+        instr_t cur_instr_decoded;
+        instr_init(GLOBAL_DCONTEXT, &cur_instr_decoded);
+        const app_pc decode_pc = const_cast<app_pc>(shard->prev_instr_.instr.encoding);
+        const app_pc next_pc =
+            decode_from_copy(GLOBAL_DCONTEXT, decode_pc,
+                             reinterpret_cast<app_pc>(trace_pc), &cur_instr_decoded);
+        if (next_pc == nullptr) {
+            instr_free(GLOBAL_DCONTEXT, &cur_instr_decoded);
```

Comment:
If decoding fails, we also want to set cur_instr_decoded to nullptr (also see other comment about making it heap-allocated).